src/Fork.ts

```diff
@@ -36,11 +36,13 @@ export const race = <const Fxs extends readonly Fx<unknown, unknown>[]>(fxs: Fxs
   return Task.race(ps)
 }) as Fx<Exclude<EffectsOf<Fxs[number]>, Async | Fail<any>> | Fork, Task.Task<ResultOf<Fxs[number]>, ErrorsOf<EffectsOf<Fxs[number]>>>>
 
-export const bounded = (maxConcurrency: number) => <const E, const A>(f: Fx<E, A>) => {
+export const bounded = (maxConcurrency: number) => <const E, const A>(f: Fx<E, A>) => fx(function* () {
+  // The HandlerContext for this bounded concurrency scope won't change, so we can cache it
+  const c = yield* getHandlerContext
   const s = new Semaphore(maxConcurrency)
-  return f.pipe(
-    handle(Fork, f => ok(acquireAndRunFork(f, s)))
+  return yield* f.pipe(
+    control(Fork, (resume, f) => ok(resume(acquireAndRunFork(f, s, c))))
```

Comment:
Since fork handlers define a kind of scope, they must not be shared/moved around, so switching to a `control` handler to prevent that.